File: Longest_Palindromic_Substring/Longest_Palindromic_Substring.cpp
#include <string>
#include <iostream>
#include <locale>

using namespace std;

//General Time complexity O(n*n)
class Solution1{
public:
    string longestPalindrome (string s){
        int startIdx = 0, left = 0, right = 0, len = 0;
        for(int i = 0; i < s.size(); ++i){
            if (s[i] == s[i + 1]){
                left = i;
                right = i + 1;
                searchPalindrome(s, left, right, startIdx, len);
            }
            left = right = i;
            searchPalindrome(s, left, right, startIdx, len);
        }
        if(len == 0)
            len = s.size();
        return s.substr(startIdx, len);
    }
    
    void searchPalindrome(string s, int left, int right, int &startIdx, int &len){
        int step = 1;
        while((left - step) >= 0 && (right + step) < s.size()){
            if (s[left - step] != s[right + step])
                break;
            ++step;
        }
        int wide = right - left + 2 * step - 1;
        if(len < wide){
            len = wide;
            startIdx = left - step + 1;
        }
    }
};

//DP
class Solution_DP{
public:
    string longestPalindrome(string s){
        int dp[s.size()][s.size()] = {0}, left = 0, right = 0, len = 0;
        for (int i = 0; i < s.size(); ++i){
            for (int j = 0; j < i; ++j){
                dp[j][i] = (s[i] == s[j] && (i - j < 2 || dp[j + 1][i - 1]));
                if(dp[j][i] && len < i - j + 1){
                    len = i - j + 1;
                    left = j;
                    right = i;
                }
            }
            dp[i][i] = 1;
        }
        return s.substr(left, right - left + 1);
    }
};

//Manacher's Algorithm
class Solution_Manacher{
public:
    string longestPalindrome(string s){
        string t = "$#";
        for(int i = 0; i < s.size(); ++i){
            t += s[i];
            t += '#';
        }
        int p[t.size()] = {0}, id = 0, mx = 0, resId = 0, resMx = 0;
        for(int i = 0; i < t.size(); ++i){
            p[i] = mx > i ? min(p[2 * id - 1], mx - i) : 1;
            while(t[i + p[i]] == t[i - p[i]]) ++p[i];
            if(mx < i + p[i]){
                mx = i + p[i];
                id = i;
            }
            if(resMx < p[i]){
                resMx = p[i];
                resId = i;
            }
        }
        return s.substr((resId - resMx) / 2, resMx - 1);
    }
};

int main()
{
    //locale::global(locale(""));
    //setlocale(LC_CTYPE, "");    // MinGW gcc.
    //wcout.imbue(locale(""));
    Solution1 *Solution = new Solution1();
    cout << "General" << endl << Solution->longestPalindrome("adhjhdf") << endl;
    Solution_DP Solution_1;
    cout << "DP" << endl << Solution_1.longestPalindrome("adhjhdf") << endl;
    Solution_Manacher Solution_2;
    cout << "Manacher" << endl << Solution_2.longestPalindrome("adhjhdf") << endl;
}

File: _site/Longest_Palindromic_Substring.py
#General Way Time complexity O(n^2)
class Solution_General:
    def searchPalindrome(self, s, left, right, startIdx, length):
        step = 1
        while((left - step) >= 0 and (right + step) < len(s)):
            if s[left - step] != s[right + step]:
                break
            step += 1
        wide = right - left + 2 * step - 1
        if length < wide:
            length = wide
            startIdx = left - step + 1
        return s, left, right, startIdx, length

    def longestPalindrome(self, s):
        startIdx = 0
        left = 0 
        right = 0 
        length = 0
        for i in range(0, len(s) - 1):
            if (s[i] == s[i + 1]):
                left = i
                right = i + 1
                s, left, right, startIdx, length = self.searchPalindrome(s, left, right, startIdx, length)
            left = right = i
            s, left, right, startIdx, length = self.searchPalindrome(s, left, right,startIdx, length)
        
        if length == 0:
            length = len(s)
        return s[startIdx : startIdx + length]

#DP Time complexity O(n^2)
class Solution_DP:
    def longestPalindrome(self, s):
        x_axis = y_axis = len(s)
        dp = [[0 for col in range(x_axis)] for row in range(y_axis)]
        left = 0
        right = 0
        length = 0
        for i in range(0, len(s)):
            for j in range(0, i):
                dp[j][i] = (s[i] == s[j] and (i - j < 2 or dp[j + 1][i - 1]))
                if(dp[j][i] and length < i - j + 1):
                    length = i - j + 1
                    left = j
                    right = i
            dp[i][i] = 1
        return s[left : right + 1]

#Manacher's Algorithm Time complexity O(n)
class Solution_Manacher:
    def longestPalindrome(self, s):
        t = '#'.join('^{}$'.format(s)) #填充首尾与分割符使得长度为奇数
        p = [0] * len(t)
        id = 0
        mx = 0
        resId = 0
        resMx = 0
        for i in range(0, len(t), 1):
            if mx > i:
                p[i] = min(p[2 * id - i], mx - i)
            else:
                p[i] = 1
            while t[i + p[i] - 1] == t[i - p[i] - 1]:
                p[i] += 1
            if mx < i + p[i]:
                mx = i + p[i]
                id = i
            if resMx < p[i]:
                resMx = p[i]
                resId = i
        return s[(resId - resMx) // 2 : (resId + resMx) // 2 - 1]

if __name__=="__main__":       
    Solution0 = Solution_General()
    Solution1 = Solution_DP()
    Solution2 = Solution_Manacher()
    print("General Way: " + Solution0.longestPalindrome("adhda"))
    print("DP: " + Solution1.longestPalindrome("adhda"))
    print("Manacher: " + Solution2.longestPalindrome("adhda"))
    